fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexer.java

```diff
@@ -223,23 +152,73 @@
                   @Nonnull IndexStatePrecondition indexStatePrecondition,
                   boolean useSynchronizedSession,
                   long leaseLengthMillis,
-                  boolean trackProgress) {
+                  boolean trackProgress,
+                  @Nonnull IndexFromIndexPolicy indexFromIndex) {
         this.runner = runner;
-        this.recordStoreBuilder = recordStoreBuilder;
         this.index = index;
-        this.recordTypes = recordTypes;
-        this.configLoader = configLoader;
-        this.config = config;
-        this.limit = config.maxLimit;
-        this.syntheticIndex = syntheticIndex;
-        this.indexStatePrecondition = indexStatePrecondition;
-        this.useSynchronizedSession = useSynchronizedSession;
-        this.leaseLengthMills = leaseLengthMillis;
-        this.trackProgress = trackProgress;
-
-        this.recordsRange = computeRecordsRange();
-        timeOfLastProgressLogMillis = System.currentTimeMillis();
-        totalRecordsScanned = new AtomicLong(0);
+        this.indexFromIndex = indexFromIndex;
+
+        this.common = new OnlineIndexerCommon(runner, recordStoreBuilder,
+                index, recordTypes, configLoader, config,
+                syntheticIndex,
+                indexStatePrecondition.getCommonVal(),
+                trackProgress,
+                useSynchronizedSession,
+                leaseLengthMillis
+            );
+    }
+
+    @Nonnull
+    private OnlineIndexerByIndex getScannerByIndex() {
+        if (! (indexScanner instanceof OnlineIndexerByIndex)) { // this covers null pointer
```

Comment:
```
// this covers null pointer
```
You mean `indexScanner` is nonnull, right?

Reply:
Yes. When writing something `if(A != null && A instanceof B)`, the IDE is grumbling about redundancy.

Reply:
Yeah, that's because `null instanceof Foo` is always `false` according to the definition of the `instanceof` operator. Note that `(Foo)null` will not throw a cast class exception, though, so `x instanceof Foo` does not return true if and only if `(Foo)x` will succeed without error, which I think is kind of confusing and probably wouldn't be what they did if they had to make Java all over again (or at least, isn't what I'd do if I had to make Java all over again)